iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/pipe/resource/ref/PipePhantomReferenceManager.java

```diff
@@ -0,0 +1,137 @@
+
+  private static final Set<PipeEventPhantomReference> PIPE_EVENT_PHANTOM_REFERENCES =
+      ConcurrentHashMap.newKeySet();
+
+  private static final ReferenceQueue<EnrichedEvent> REFERENCE_QUEUE = new ReferenceQueue<>();
+
+  public PipePhantomReferenceManager() {
+    // Do nothing now.
+  }
+
+  public int getPhantomReferenceCount() {
+    return PIPE_EVENT_PHANTOM_REFERENCES.size();
+  }
+
+  protected void gcHook() {
+    Reference<? extends EnrichedEvent> reference;
+    try {
+      while ((reference = REFERENCE_QUEUE.remove(500)) != null) {
+        finalizeResource((PipeEventPhantomReference) reference);
+      }
+    } catch (final InterruptedException e) {
+      // Finalize remaining references.
+      while ((reference = REFERENCE_QUEUE.poll()) != null) {
+        finalizeResource((PipeEventPhantomReference) reference);
+      }
+    } catch (final Exception e) {
+      // Nowhere to really log this.
+    }
+  }
+
+  private void finalizeResource(final PipeEventPhantomReference reference) {
+    try {
+      reference.finalizeResources();
+      reference.clear();
+    } finally {
+      PIPE_EVENT_PHANTOM_REFERENCES.remove(reference);
+    }
+  }
+
+  private static class PipeEventPhantomReference extends PhantomReference<EnrichedEvent> {
+
+    private final String coreReportMessageSnapshot;
+    private PipeEventResource resource;
+
+    private PipeEventPhantomReference(
+        final EnrichedEvent event,
+        final PipeEventResource resource,
+        final ReferenceQueue<? super EnrichedEvent> queue) {
+      super(event, queue);
+      this.coreReportMessageSnapshot = event.coreReportMessage();
+      this.resource = resource;
+    }
+
+    private void finalizeResources() {
+      if (this.resource != null) {
+        try {
+          this.resource.clearReferenceCount(coreReportMessageSnapshot);
+        } finally {
+          this.resource = null;
+        }
+      }
+    }
+  }
+
+  ///////////////////// APIs provided for EnrichedEvent /////////////////////
+
+  public void trackPipeEventResource(final EnrichedEvent event, final PipeEventResource resource) {
+    final PipeEventPhantomReference reference =
+        new PipeEventPhantomReference(event, resource, REFERENCE_QUEUE);
+    PIPE_EVENT_PHANTOM_REFERENCES.add(reference);
+  }
+
+  public abstract static class PipeEventResource {
+
+    private final AtomicBoolean isReleased;
+    private final AtomicInteger referenceCount;
+
+    protected PipeEventResource(
+        final AtomicBoolean isReleased, final AtomicInteger referenceCount) {
+      this.isReleased = isReleased;
+      this.referenceCount = referenceCount;
+    }
+
+    private void clearReferenceCount(final String holderMessage) {
+      if (isReleased.get()) {
+        return;
+      }
+
+      if (referenceCount.get() >= 1) {
+        LOGGER.info("finalize resource for event with holder message {}", holderMessage);
```

Comment:
```suggestion
        LOGGER.error("finalize resource for event with holder message {}", holderMessage);
```
"RESOURCE LEAK DETECTED: {}....", class name...